kotlinx-coroutines-core/jvm/src/internal/FastServiceLoader.kt

```diff
@@ -20,7 +27,59 @@ import java.util.zip.*
 internal object FastServiceLoader {
     private const val PREFIX: String = "META-INF/services/"
 
-    internal fun <S> load(service: Class<S>, loader: ClassLoader): List<S> {
+    /**
+     * This method attempts to load [MainDispatcherFactory] in Android-friendly way.
+     *
+     * If we are not on Android, this method fallbacks to a regular service loading,
+     * else we attempt to do `Class.forName` lookup for
+     * `AndroidDispatcherFactory` and `TestMainDispatcherFactory`.
+     * If lookups are successful, we return resulting instances because we know that
+     * `MainDispatcherFactory` API is internal and this is the only possible classes of `MainDispatcherFactory` Service on Android.
+     *
+     * Such intricate dance is required to avoid calls to `ServiceLoader.load` for multiple reasons:
+     * 1) It eliminates disk lookup on potentially slow devices on the Main thread.
+     * 2) Various Android toolchain versions by various vendors don't tend to handle ServiceLoader calls properly.
+     *    Sometimes META-INF is removed from the resulting APK, sometimes class names are mangled, etc.
+     *    While it is not the problem of `kotlinx.coroutines`, it significantly worsens user experience, thus we are workarounding it.
+     *    Examples of such issues are #932, #1072, #1557, #1567
+     *
+     * We also use SL for [CoroutineExceptionHandler], but we do not experience the same problems and CEH is a public API
+     * that may already be injected vis SL, so we are not using the same technique for it.
+     */
+    internal fun loadMainDispatcherFactory(): List<MainDispatcherFactory> {
+        val clz = MainDispatcherFactory::class.java
+        if (!ANDROID_DETECTED) {
+            return load(clz, clz.classLoader)
+        }
+
+        return try {
+            val result = ArrayList<MainDispatcherFactory>()
+            createInstanceOf(clz, "kotlinx.coroutines.android.AndroidDispatcherFactory")?.apply { result.add(this) }
+            createInstanceOf(clz, "kotlinx.coroutines.test.internal.TestMainDispatcherFactory")?.apply { result.add(this) }
+            result
+        } catch (e: Throwable) {
+            // Fallback to the regular SL in case of any unexpected exception
+            load(clz, clz.classLoader)
+        }
+    }
+
+    /*
+     * This method is inline to have a direct Class.forName("string literal") in the byte code to avoid weird interactions with ProGuard/R8.
+     */
+    @Suppress("NOTHING_TO_INLINE")
+    private inline fun createInstanceOf(
+        baseClass: Class<MainDispatcherFactory>,
+        serviceClass: String
+    ): MainDispatcherFactory? {
+        return try {
+            val clz = Class.forName(serviceClass, true, baseClass.classLoader)
```

Comment:
We should confirm that calls to `forName` in this three-argument form can be replaced with class constants by R8. I'll do that.

Reply:
Thanks!

Reply:
So far found at least one problem: https://issuetracker.google.com/issues/141678220

The use of the three-arg variant of this method also means that R8 cannot (currently) rewrite this into a `const-class` bytecode at compile time. This is probably okay, since this code only runs once. And it's something that can be investigated in the future. I actually hope to get code like this eventually rewritten as a direct call to the constructor!